src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -1538,84 +1692,65 @@ DistributionColumnUsesGeneratedStoredColumn(TupleDesc relationDesc,
  * be dropped.
  */
 void
-UndistributeTable(Oid relationId)
+ConvertTable(char conversionType, Oid relationId, char *distributionColumn, int shardCount, char *colocateWith, char *accessMethod, bool cascadeToColocated)
 {
-	Relation relation = try_relation_open(relationId, ExclusiveLock);
-	if (relation == NULL)
-	{
-		ereport(ERROR, (errmsg("cannot undistribute table"),
-						errdetail("because no such distributed table exists")));
-	}
-
-	relation_close(relation, NoLock);
-
-	if (!IsCitusTable(relationId))
-	{
-		ereport(ERROR, (errmsg("cannot undistribute table "),
-						errdetail("because the table is not distributed")));
-	}
-
-	if (TableReferencing(relationId))
-	{
-		ereport(ERROR, (errmsg("cannot undistribute table "
-							   "because it has a foreign key")));
-	}
-
-	if (TableReferenced(relationId))
-	{
-		ereport(ERROR, (errmsg("cannot undistribute table "
-							   "because a foreign key references to it")));
-	}
-
-	char relationKind = get_rel_relkind(relationId);
-	if (relationKind == RELKIND_FOREIGN_TABLE)
+	List *colocatedTableList = NIL;
+	if (cascadeToColocated)
 	{
-		ereport(ERROR, (errmsg("cannot undistribute table "
-							   "because it is a foreign table")));
+		colocatedTableList = ColocatedTableList(relationId);
 	}
 
-	if (PartitionTable(relationId))
+	bool shardCountIsNull = false;
+	if (shardCount == 0)
 	{
-		Oid parentRelationId = PartitionParentOid(relationId);
-		char *parentRelationName = get_rel_name(parentRelationId);
-		ereport(ERROR, (errmsg("cannot undistribute table "
-							   "because it is a partition"),
-						errhint("undistribute the partitioned table \"%s\" instead",
-								parentRelationName)));
+		CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(relationId);
+		shardCount = cacheEntry -> shardIntervalArrayLength;
+		shardCountIsNull = true;
 	}
-
-	List *preLoadCommands = GetPreLoadTableCreationCommands(relationId, true);
+	List *preLoadCommands = GetPreLoadTableCreationCommands(relationId, true, accessMethod);
 	List *postLoadCommands = GetPostLoadTableCreationCommands(relationId);
 
 	postLoadCommands = list_concat(postLoadCommands,
 								   GetViewCreationCommandsOfTable(relationId));
 
+	char *relationName = get_rel_name(relationId);
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+
 	int spiResult = SPI_connect();
 	if (spiResult != SPI_OK_CONNECT)
 	{
 		ereport(ERROR, (errmsg("could not connect to SPI manager")));
 	}
 
-	char *relationName = get_rel_name(relationId);
-	Oid schemaId = get_rel_namespace(relationId);
-	char *schemaName = get_namespace_name(schemaId);
+	bool isPartitionTable = false;
+	char *attachToParentCommand = NULL;
+	if (PartitionTable(relationId))
+	{
+		isPartitionTable = true;
+		char *detachFromParentCommand = GenerateDetachPartitionCommand(relationId);
+		attachToParentCommand = GenerateAlterTableAttachPartitionCommand(relationId);
+
+		spiResult = SPI_execute(detachFromParentCommand, false, 0);
```

Comment:
it would be nice to do this at the last moment (just before drop or attach), to avoid blocking reads from the beginning